substrate/primitives/runtime/src/traits/transaction_extension/mod.rs

```diff
@@ -0,0 +1,526 @@
+/// [validate](TransactionExtension::validate) function is fairly sophisticated and warrants
+/// further explanation.
+///
+/// Firstly, the `origin` parameter. The `origin` passed into the first item in a pipeline is simply
+/// that passed into the tuple itself. It represents an authority who has authorized the implication
+/// of the transaction, as of the extension it has been passed into *and any further extensions it
+/// may pass though, all the way to, and including, the transaction's dispatch call itself. Each
+/// following item in the pipeline is passed the origin which the previous item returned. The origin
+/// returned from the final item in the pipeline is the origin which is returned by the tuple
+/// itself.
+///
+/// This means that if a constituent extension returns a different origin to the one it was called
+/// with, then (assuming no other extension changes it further) *this new origin will be used for
+/// all extensions following it in the pipeline, and will be returned from the pipeline to be used
+/// as the origin for the call's dispatch*. The call itself as well as all these extensions
+/// following may each imply consequence for this origin. We call this the *inherited implication*.
+///
+/// The *inherited implication* is the cumulated on-chain effects born by whatever origin is
+/// returned. It is expressed to the [validate](TransactionExtension::validate) function only as the
+/// `implication` argument which implements the [Encode] trait. A transaction extension may define
+/// its own implications through its own fields and the
+/// [implicit](TransactionExtensionBase::implicit) function. This is only utilized by extensions
+/// which preceed it in a pipeline or, if the transaction is an old-school signed trasnaction, the
+/// underlying transaction verification logic.
+///
+/// **The inherited implication passed as the `implication` parameter to
+/// [validate](TransactionExtension::validate) does not include the extension's inner data itself
+/// nor does it include the result of the extension's `implicit` function.** If you both provide an
+/// implication and rely on the implication, then you need to manually aggregate your extensions
+/// implication with the aggregated implication passed in.
+pub trait TransactionExtension<Call: Dispatchable, Context>: TransactionExtensionBase {
+	/// The type that encodes information that can be passed from validate to prepare.
+	type Val;
+
+	/// The type that encodes information that can be passed from prepare to post-dispatch.
+	type Pre;
+
+	/// Validate a transaction for the transaction queue.
+	///
+	/// This function can be called frequently by the transaction queue to obtain transaction
+	/// validity against current state. It should perform all checks that determine a valid
+	/// transaction, that can pay for its execution and quickly eliminate ones that are stale or
```

Comment:
This is called for all kinds of transactions, also `General` transactions which do not really pay for their execution.

Reply:
`General` transactions indeed can't pay for their execution as things stand now. These transactions would either need to be feeless with some other means of approving/validating the origin to prevent spam or people who use this `General` format would need to write their own logic for a transaction payment extension.

Reply:
https://github.com/paritytech/polkadot-sdk/issues/2415#issuecomment-1989480626